packages/vhd-lib/Vhd/VhdDirectory.test.js

```diff
@@ -0,0 +1,126 @@
+    await convertToVhdDirectory(childRawDirectoryName, childDirectoryFileName, childDirectoryName)
+
+    await Disposable.use(async function* () {
+      const parentVhd = yield openVhd(handler, 'parent.dir.vhd', { flags: 'w' })
+      await parentVhd.readBlockAllocationTable()
+      const childFileVhd = yield openVhd(handler, 'childFile.vhd')
+      await childFileVhd.readBlockAllocationTable()
+      const childDirectoryVhd = yield openVhd(handler, 'childDir.vhd')
+      await childDirectoryVhd.readBlockAllocationTable()
+
+      let childBlockData = (await childDirectoryVhd.readBlock(0)).data
+      await parentVhd.mergeBlock(childDirectoryVhd, 0)
+      await parentVhd.writeFooter()
+      await parentVhd.writeBlockAllocationTable()
+      let parentBlockData = (await parentVhd.readBlock(0)).data
+      // block should be present in parent
+      assert.equal(parentBlockData.equals(childBlockData), true)
+      // block should not be in child since it's a rename  for vhd directory
+      await assert.rejects(childDirectoryVhd.readBlock(0))
+
+      childBlockData = (await childFileVhd.readBlock(1)).data
+      await parentVhd.mergeBlock(childFileVhd, 1)
+      await parentVhd.writeFooter()
+      await parentVhd.writeBlockAllocationTable()
+      parentBlockData = (await parentVhd.readBlock(1)).data
+      // block should be present in parent in case of mixed vhdfile/vhddirectory
+      assert.equal(parentBlockData.equals(childBlockData), true)
+      // block should still be child
+      await childFileVhd.readBlock(1)
+    })
+  })
+
+  it('compresses blocks and metadata works', async () => {
+    const initalSize = 4
+    const rawFileName = `${tempDir}/randomfile`
+    const vhdName = `${tempDir}/parent.vhd`
+
+    await createRandomFile(rawFileName, initalSize)
+    await convertFromRawToVhd(rawFileName, vhdName)
+    await Disposable.use(async function* () {
+      const vhd = yield openVhd(handler, 'parent.vhd')
+      await vhd.readBlockAllocationTable()
+      const compressedVhd = yield VhdDirectory.create(handler, 'compressed.vhd', { compression: 'gzip' })
+      compressedVhd.header = vhd.header
+      compressedVhd.footer = vhd.footer
+      for await (const block of vhd.blocks()) {
+        await compressedVhd.writeEntireBlock(block)
+      }
+      await Promise
+        .all[(await compressedVhd.writeHeader(), await compressedVhd.writeFooter(), await compressedVhd.writeBlockAllocationTable())]
+
+      // compressed vhd have a metadata file
+      assert.equal(await fs.exists(`${tempDir}/compressed.vhd/chunk-filters.json`), true)
+      const metada = JSON.parse(await handler.readFile('compressed.vhd/chunk-filters.json'))
+      assert.equal(metada[0], 'gzip')
+
+      // compressed vhd should not be broken
+      await compressedVhd.readHeaderAndFooter()
+      await compressedVhd.readBlockAllocationTable()
+
+      // check that footer and header are not modified
+      assert.deepEqual(compressedVhd.footer, vhd.footer)
```

Comment:
I am not sure it can comapre buffer correctly, that why I was using `buffer1.equals(buffer2) === true`
I will test it tomorrow to be sure

Reply:
Were you able to try? I hadn't noticed any error during my test.

Reply:
tested right now, it works and doesn't crash even on big buffers